In [2]:
import math
from geometry_msgs.msg import Pose
from tf_transformations import quaternion_from_euler

def euler_to_quaternion(roll_deg, pitch_deg, yaw_deg):
    """
    将欧拉角（度）转换为四元数。
    
    :param roll_deg: 滚转角（度）
    :param pitch_deg: 俯仰角（度）
    :param yaw_deg: 偏航角（度）
    :return: 四元数 (x, y, z, w)
    """
    # 将角度转换为弧度
    roll = math.radians(roll_deg)
    pitch = math.radians(pitch_deg)
    yaw = math.radians(yaw_deg)
    
    # 使用 tf_transformations 库将欧拉角转换为四元数
    q = quaternion_from_euler(roll, pitch, yaw)
    
    return q  # 返回四元数 (x, y, z, w)

def main():
    # 示例欧拉角（以度为单位）
    roll_deg = 90.0   # 滚转角
    pitch_deg = 90.0  # 俯仰角
    yaw_deg = 0.0    # 偏航角
    
    # 转换为四元数
    quaternion = euler_to_quaternion(roll_deg, pitch_deg, yaw_deg)
    
    # 创建一个 Pose 消息并赋值
    target_pose = Pose()
    target_pose.orientation.x = quaternion[0]
    target_pose.orientation.y = quaternion[1]
    target_pose.orientation.z = quaternion[2]
    target_pose.orientation.w = quaternion[3]
    
    # 打印结果
    print("四元数:")
    print(f"x: {target_pose.orientation.x}")
    print(f"y: {target_pose.orientation.y}")
    print(f"z: {target_pose.orientation.z}")
    print(f"w: {target_pose.orientation.w}")

if __name__ == "__main__":
    main()


四元数:
x: 0.5
y: 0.5
z: -0.4999999999999999
w: 0.5000000000000001


In [5]:
# 输入xy位置获取 roll_deg, pitch_deg, yaw_deg
y = 0.946
x = 0.245
yaw = math.atan(y/x)
print(yaw/3.1415926*180)
print(x*0.76, y*0.76)

75.4802606087377
0.1862 0.7189599999999999


In [9]:
# 输入 关节角的度数，得到弧度值  egp64的位置

joint = [-26, -36, -143, 0, -91, 64 ]
rad = []
for i in range(len(joint)):
    rad_value = joint[i]*3.1415926/180
    rad.append(format(rad_value, ".4f"))
print(rad)

['-0.4538', '-0.6283', '-2.4958', '0.0000', '-1.5882', '1.1170']


### 移动到egp6的4位置

    {
      "action_name": "move_arm_joint",
      "parameters": {
        "joints_value": {
          "type": "JointPathParameter",
          "joint_0": -0.4538,
          "joint_1": -0.6283,
          "joint_2": -2.4958,
          "joint_3": 0.0,
          "joint_4": -1.5882,
          "joint_5": 1.1170
        },
        "constrain": {
          "type": "ConstrainParameter",
          "is_constrain": false,
          "x_axis_tolerance": 0.1,
          "y_axis_tolerance": 0.1,
          "z_axis_tolerance": 0.1
        }
      }
    }

In [10]:
# 输入 关节角的度数，得到弧度值  egp64的位置

joint = [-10 ,-29, -154, 0, -87, 80 ]
rad = []
for i in range(len(joint)):
    rad_value = joint[i]*3.1415926/180
    rad.append(format(rad_value, ".4f"))
print(rad)

['-0.1745', '-0.5061', '-2.6878', '0.0000', '-1.5184', '1.3963']


### 移动到pipettle的位置    

    {
      "action_name": "move_arm_joint",
      "parameters": {
        "joints_value": {
          "type": "JointPathParameter",
          "joint_0": -0.175,
          "joint_1": -0.5061,
          "joint_2": -2.6878,
          "joint_3": 0.0,
          "joint_4": -1.5184,
          "joint_5": 1.3963
        },
        "constrain": {
          "type": "ConstrainParameter",
          "is_constrain": false,
          "x_axis_tolerance": 0.1,
          "y_axis_tolerance": 0.1,
          "z_axis_tolerance": 0.1
        }
      }
    }

In [1]:
# 输入 关节角的度数，得到弧度值  放回egp64的位置

joint = [-35 ,-12, -158, 0, -100, 55 ]
rad = []
for i in range(len(joint)):
    rad_value = joint[i]*3.1415926/180
    rad.append(format(rad_value, ".4f"))
print(rad)

['-0.6109', '-0.2094', '-2.7576', '0.0000', '-1.7453', '0.9599']


In [8]:
# 输入 关节角的度数，得到弧度值  到达瓶子的位置

joint = [76 , 46, 13, -179, 59, 89 ]
rad = []
for i in range(len(joint)):
    rad_value = joint[i]*3.1415926/180
    rad.append(format(rad_value, ".4f"))
print(rad)

['1.3265', '0.8029', '0.2269', '-3.1241', '1.0297', '1.5533']


In [5]:
# 输入 关节角的度数，得到弧度值  到达开瓶器的位置

joint = [0.5 ,41, 30, -180, 72, 90 ]
rad = []
for i in range(len(joint)):
    rad_value = joint[i]*3.1415926/180
    rad.append(format(rad_value, ".4f"))
print(rad)

['0.0087', '0.7156', '0.5236', '-3.1416', '1.2566', '1.5708']


In [6]:
# 输入 关节角的度数，得到弧度值  到达pipettle的位置

joint = [-11 ,-27, -157, 0, -87, 80]
rad = []
for i in range(len(joint)):
    rad_value = joint[i]*3.1415926/180
    rad.append(format(rad_value, ".4f"))
print(rad)

['-0.1920', '-0.4712', '-2.7402', '0.0000', '-1.5184', '1.3963']


In [7]:
# 输入 关节角的度数，得到弧度值  移液器到达瓶子上方

joint = [74 ,50, -26, 0, 65, -106]
rad = []
for i in range(len(joint)):
    rad_value = joint[i]*3.1415926/180
    rad.append(format(rad_value, ".4f"))
print(rad)

['1.2915', '0.8727', '-0.4538', '0.0000', '1.1345', '-1.8500']


In [9]:
# 输入 关节角的度数，得到弧度值  移液器要返回的位置

joint = [-16 ,0, -165, 0, -106, 75]
rad = []
for i in range(len(joint)):
    rad_value = joint[i]*3.1415926/180
    rad.append(format(rad_value, ".4f"))
print(rad)

['-0.2793', '0.0000', '-2.8798', '0.0000', '-1.8500', '1.3090']


In [3]:
# 输入 关节角的度数，得到弧度值  移液器要到达试管的位置

joint = [120 ,31, -17, 0, 77, -58]
rad = []
for i in range(len(joint)):
    rad_value = joint[i]*3.1415926/180
    rad.append(format(rad_value, ".4f"))
print(rad)

['2.0944', '0.5411', '-0.2967', '0.0000', '1.3439', '-1.0123']


In [9]:
# 输入 关节角的度数，得到弧度值  egp64要到达试管的位置

joint = [-62 ,-17, -186, 4, 26, 92]
rad = []
for i in range(len(joint)):
    rad_value = joint[i]*3.1415926/180
    rad.append(format(rad_value, ".4f"))
print(rad)

['-1.0821', '-0.2967', '-3.2463', '0.0698', '0.4538', '1.6057']


In [10]:
# 输入 关节角的度数，得到弧度值  egp64要到达核磁仪的位置

joint = [98 ,-1, -152, 175, 25, -80]
rad = []
for i in range(len(joint)):
    rad_value = joint[i]*3.1415926/180
    rad.append(format(rad_value, ".4f"))
print(rad)

['1.7104', '-0.0175', '-2.6529', '3.0543', '0.4363', '-1.3963']


In [16]:
# 输入 关节角的度数，得到弧度值  放回试管的位置

joint = [-65 ,-7, -172, 158, -2, -63]
rad = []
for i in range(len(joint)):
    rad_value = joint[i]*3.1415926/180
    rad.append(format(rad_value, ".4f"))
print(rad)

['-1.1345', '-0.1222', '-3.0020', '2.7576', '-0.0349', '-1.0996']


In [13]:
# 输入 关节角的度数，得到弧度值  放置到垃圾桶的位置

joint = [50 ,-1, -152, 175, 25, -80]
rad = []
for i in range(len(joint)):
    rad_value = joint[i]*3.1415926/180
    rad.append(format(rad_value, ".4f"))
print(rad)

['0.8727', '-0.0175', '-2.6529', '3.0543', '0.4363', '-1.3963']


In [15]:
# 输入 关节角的度数，得到弧度值  倾倒液体

joint = [50 ,-1, -152, 175, 25, -200]
rad = []
for i in range(len(joint)):
    rad_value = joint[i]*3.1415926/180
    rad.append(format(rad_value, ".4f"))
print(rad)

['0.8727', '-0.0175', '-2.6529', '3.0543', '0.4363', '-3.4907']
